doozerlib/metadata.py

```diff
@@ -190,37 +294,109 @@ def cgit_url(self, filename: str, commit_hash: Optional[str] = None, branch: Opt
-            return builds[0]
-
-    def get_latest_build_info(self, default=-1):
+            package_info = koji_api.getPackage(package_name)  # e.g. {'id': 66873, 'name': 'atomic-openshift-descheduler-container'}
+            if not package_info:
+                raise IOError(f'No brew package is defined for {package_name}')
+            package_id = package_info['id']  # we could just constrain package name using pattern glob, but providing package ID # should be a much more efficient DB query.
+            # listBuilds returns all builds for the package; We need to limit the query to the builds
+            # relevant for our major/minor.
+
+            # RPMs do not have a 'v' in front of their version; images do.
+            if self.meta_type == 'image':
+                ver_prefix = 'v'  # openshift-enterprise-console-container-v4.7.0-202106032231.p0.git.d9f4379
+            else:
+                ver_prefix = ''  # openshift-clients-4.7.0-202106032231.p0.git.e29b355.el8
+
+            pattern_prefix = f'{package_name}-{ver_prefix}{self.branch_major_minor()}.'
+
+            if assembly is None:
+                assembly = self.runtime.assembly
+
+            def latest_build_list(pattern_suffix):
+                # We need '*' after pattern_suffix for RPMs, which have an extra .el7/.el8 on their
+                # release field, normally.
+                builds = koji_api.listBuilds(packageID=package_id,
+                                             state=None if build_state is None else build_state.value,
+                                             pattern=f'{pattern_prefix}{extra_pattern}{pattern_suffix}*',
+                                             queryOpts={'limit': 1, 'order': '-creation_event_id'})
+
+                # Ensure the suffix ends the string OR at least terminated by a '.' .
+                # This latter check ensures that 'assembly.how' doesn't not match a build from
+                # "assembly.howdy'.
+                refined = [b for b in builds if b['nvr'].endswith(pattern_suffix) or f'{pattern_suffix}.' in b['nvr']]
+
+                if refined:
+                    # A final sanity check to see if the build is tagged with something we
+                    # respect. There is a chance that a human may untag a build. There
+                    # is no standard practice at present in which they should (they should just trigger
+                    # a rebuild). If we find the latest build is not tagged appropriately, blow up
+                    # and let a human figure out what happened.
+                    check_nvr = refined[0]['nvr']
+                    tags = koji_api.listTags(build=check_nvr, pattern=f'{self.branch()}*')
+                    if not tags:
+                        raise IOError(f'Expected to find tag starting with {self.branch()} on latest build {check_nvr}; something has changed tags in an unexpected way')
```

Comment:
Here, if `self.branch()` is `-el7` and the `check_nvr` is a `-el8` build, this will panic:

```python
Traceback (most recent call last):
  File "/workspaces/doozer/./doozerlib/runtime.py", line 99, in wrapper
    return func(*args, **kwargs)
  File "/workspaces/doozer/./doozerlib/runtime.py", line 111, in wrapper
    return func(*args)
  File "/workspaces/doozer/./doozerlib/runtime.py", line 1409, in <lambda>
    lambda meta, _: (meta, meta.needs_rebuild()),
  File "/workspaces/doozer/./doozerlib/metadata.py", line 430, in needs_rebuild
    latest_build = self.get_latest_build(default=None)
  File "/workspaces/doozer/./doozerlib/metadata.py", line 373, in get_latest_build
    builds = latest_build_list('.assembly.stream')
  File "/workspaces/doozer/./doozerlib/metadata.py", line 361, in latest_build_list
    raise IOError(f'Expected to find tag starting with {self.branch()} on latest build {check_nvr}; something has changed tags in an unexpected way')
OSError: Expected to find tag starting with rhaos-4.9-rhel-7 on latest build openshift-4.9.0-202106031118.p0.git.27ac81c.assembly.stream.el8; something has changed tags in an unexpected way
```